ci(root): Create New Cloud Deployment

Cliftonz opened the pull request:
This creates a new pipeline for getting novu deployed quicker to the cloud

Cliftonz committed:
temp

Cliftonz committed:
Merge branch 'refs/heads/next' into inf-359-create-a-new-cloud-deployment-pipeline-for-novu-cloud

Cliftonz committed:
Refactor GitHub actions for deployment process

The GitHub actions for the deployment process has been refactored. This included renaming the deployment steps, restructuring the workflow, and moving related files into the cloud-deploy directory. Each part of the application now has a dedicated release step for better structure and maintainability.

Cliftonz committed:
Add failure and success notifications to cloud deployment workflow

This update enhances the cloud deployment workflow by adding provisions for both successful and failed release notifications. Now, if any of the previous jobs fail, a 'Release has failed' message will be posted on the '#eng-feed-deployments' Slack channel. Likewise, if all previous jobs succeed, a success notification will be sent to the same channel.

rifont commented on .github/workflows/cloud-deploy.yml:
```diff
@@ -0,0 +1,105 @@
+        env:
+          SLACK_BOT_TOKEN: ${{ secrets.SLACK_BOT_TOKEN }}
+
+  release-api:
+    needs: [pre-release]
+    runs-on: ubuntu-latest
+    steps:
+      - uses: ./.github/workflows/cloud-deploy/cloud-deploy-api.yml
+
+  release-embed:
+    needs: [pre-release]
+    runs-on: ubuntu-latest
+    steps:
+      - uses: ./.github/workflows/cloud-deploy/cloud-deploy-api.yml
+
```
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
❓ **question:** Should these `jobs` have different arguments somewhere? They appear to be identical in their function‏

Cliftonz replied:
Your right

rifont commented on .github/workflows/cloud-deploy/cloud-deploy-api.yml:
```diff
@@ -0,0 +1,139 @@
+
+  deploy_prod_api_us:
+    needs:
+#      - deploy_prod_api_eu
```
<!-- Chores are simple tasks that must be done before the subject can be “officially” accepted. Usually, these comments reference some common process. Try to leave a link to the process description so that the reader knows how to resolve the chore.  -->
☑ **chore:** Remove the commented code‏

rifont commented on .github/workflows/cloud-deploy/cloud-deploy-api.yml:
```diff
@@ -0,0 +1,139 @@
+      environment: Production
+      service_name: api
+      terraform_workspace: novu-prod
+      # This is a workaround to an issue with matrix outputs
```
<!-- Suggestions propose improvements to the current subject. It's important to be explicit and clear on what is being suggested and why it is an improvement. Consider using patches and the blocking or non-blocking decorations to further communicate your intent.  -->
✏ **suggestion:** Could we add a small description of what this works around?‏

rifont commented on .github/workflows/cloud-deploy/cloud-deploy-worker.yml:
```diff
@@ -0,0 +1,137 @@
+    # The type of runner that the job will run on
+    runs-on: ubuntu-latest
+    needs: test_worker
+    timeout-minutes: 80
```
<!-- Issues highlight specific problems with the subject under review. These problems can be user-facing or behind the scenes. It is strongly recommended to pair this comment with a suggestion. If you are not sure if a problem exists or not, consider leaving a question.  -->
⚠ **issue:** Can we time out of this job faster? 80 could hang the merge queue for some time.‏
```suggestion
    timeout-minutes: 15
```

rifont replied:
There are 5 occurrences to consider in the PR.

Cliftonz committed:
Update deployment workflows in GitHub Actions

The cloud deployment workflows are updated in the GitHub Actions configuration to correctly match each workflow with its corresponding deployment script. This includes updates for 'embed', 'web', 'web-components', 'webhook', 'widget', 'worker', and 'ws' workflows.

Cliftonz committed:
Update deployment workflows and reduce timeout

Reduced the timeout for several workflow runs from 80 to 15 minutes. The dependency on test jobs before deployment was also removed in various workflows to streamline the deployment process. These changes speed up the deployment process and make it more efficient.

rifont commented on .github/workflows/cloud-deploy/cloud-deploy-api.yml:
```diff
@@ -0,0 +1,139 @@
+          submodules: ${{ contains (matrix.name,'-ee') }}
+
+      - name: build api
+        run: pnpm build:api --skip-nx-cache
```
<!-- Suggestions propose improvements to the current subject. It's important to be explicit and clear on what is being suggested and why it is an improvement. Consider using patches and the blocking or non-blocking decorations to further communicate your intent.  -->
✏ **suggestion:** Can we remove the skip on NX Cache for `pnpm build:api` and `pnpm build:worker`? We can benefit greatly in build speed via cache usage from the pre-deploy phase‏

rifont replied:
```suggestion
        run: pnpm build:api
```

rifont commented on .github/workflows/cloud-deploy/cloud-deploy-worker.yml:
```diff
@@ -0,0 +1,137 @@
+          submodules: ${{ contains (matrix.name,'-ee') }}
+
+      - name: build worker
+        run: pnpm build:worker --skip-nx-cache
```
Ditto, remove NX cache skip
```suggestion
        run: pnpm build:worker
```

rifont reviewed (approved):
🚀🚢⛵🛳️🚣‍♂️🛶🚤⛴️🛥️🚀

Cliftonz committed:
Remove --skip-nx-cache option from build commands

The --skip-nx-cache option has been removed from the pnpm build calls in the 'cloud-deploy-api' and 'cloud-deploy-worker' workflows. This change assumes that the tool will use its default configuration, which includes leveraging cache to improve build time.

Cliftonz committed:
Merge branch 'next' into inf-359-create-a-new-cloud-deployment-pipeline-for-novu-cloud
